lib/storage-client.js

```diff
@@ -0,0 +1,352 @@
+  }
+  return node.firstChild.nodeValue;
+}
+
+
+class ChromedriverStorageClient {
+  constructor (args = {}) {
+    const {
+      chromedriverDir = getChromedriverDir(),
+      timeout = TIMEOUT,
+    } = args;
+    this.chromedriverDir = chromedriverDir;
+    this.timeout = timeout;
+    this.mapping = {};
+  }
+
+  parseNotes (content) {
+    const versionMatch = /^\s*[-]+ChromeDriver[\D]+([\d.]+)/im.exec(content);
+    if (!versionMatch) {
+      return {};
+    }
+    const minBrowserVersionMatch = /^\s*Supports Chrome[\D]+(\d+)/im.exec(content);
+    if (!minBrowserVersionMatch) {
+      return {
+        version: versionMatch[1],
+      };
+    }
+    return {
+      version: versionMatch[1],
+      minBrowserVersion: minBrowserVersionMatch[1],
+    };
+  }
+
+  async retrieveAdditionalDriverInfo (driverKey, url, infoDict) {
+    const response = await requestPromise({
+      url,
+      method: 'GET',
+      headers: {
+        'user-agent': 'appium',
+        accept: '*/*',
+      },
+      resolveWithFullResponse: true,
+      timeout: this.timeout,
+    });
+    const { minBrowserVersion } = this.parseNotes(response.body);
+    if (!minBrowserVersion) {
+      log.debug(`The driver '${driverKey}' does not contain valid notes. Skipping it`);
+      return;
+    }
+    infoDict.minBrowserVersion = minBrowserVersion;
+  }
+
+  async parseStorageXml (doc, parseNotes = true) {
+    const driverNodes = xpath.select(`//*[local-name(.)='Contents']`, doc);
+    log.debug(`Parsed ${driverNodes.length} entries from storage XML`);
+    if (_.isEmpty(driverNodes)) {
+      return;
+    }
+
+    const fibers = [];
+    for (const driverNode of driverNodes) {
+      const key = extractNodeText(findChildNode(driverNode, 'Key'));
+      if (!_.includes(key, '/chromedriver_')) {
+        continue;
+      }
+
+      log.debug(`Processing chromedriver entry '${key}'`);
+      const etag = extractNodeText(findChildNode(driverNode, 'ETag'));
+      if (!etag) {
+        log.debug(`The entry '${key}' does not contain the checksum. Skipping it`);
+        continue;
+      }
+
+      const cdInfo = {
+        url: `${STORAGE_URL}/${key}`,
+        etag: _.trim(etag, '"'),
+        version: _.first(key.split('/')),
+      };
+      this.mapping[key] = cdInfo;
+      log.debug(`Added a new entry to the storage mapping: ${JSON.stringify(cdInfo)}`);
+
+      const notesPath = `${cdInfo.version}/notes.txt`;
+      const isNotesPresent = !!driverNodes
+        .reduce((acc, node) => acc || findChildNode(node, 'Key', notesPath), false);
+      if (!isNotesPresent) {
+        cdInfo.minBrowserVersion = null;
+        if (parseNotes) {
+          log.info(`The entry '${key}' does not contain any notes. Skipping it`);
+        }
+        continue;
+      } else if (!parseNotes) {
+        continue;
+      }
+
+      fibers.push(this.retrieveAdditionalDriverInfo(key, `${STORAGE_URL}/${notesPath}`, cdInfo));
+      if (fibers.length % MAX_PARALLEL_THREADS === 0) {
```

Comment:
What do you think about slicing the driverNodes by MAX_PARALLEL_THREADS and delegating them to a method for parallelisation? If we do that we don't need to care about modifying the fibers list and don't need to call `_.pullAll(fibers)` to empty the fibers object

Reply:
actually, pullAll is not really needed here, since B.all would anyway ignore already resolved promises